### PR TITLE
Deploy again feature

### LINKED ISSUE #168: Add button to retry a failed deploy
Sometimes you simply want to try to run the same deploy again with exactly the same parameters. There should be a button that allows you to do this on a failed deploy page.

### PR DESCRIPTION
**Pull request by jennysivapalan**
Adds a Deploy Again button to the /deploy/view/:uuid view (Fixes #168). This will add a new deploy (new uuid) with the same parameters as the deploy in view.

Some notes: 
* Button is present for both failed and successful builds
* As we don't seem to deploy to specific hosts anymore I haven't included this in parameters to pass to the form. @sihil should I?

screen shot 2016-05-27 at 16 14 14

cc/ @sihil @alexduf

### CODE REVIEW

**Comment by alexduf** [comment]
I can't really judge the PR (though it looks good), but the idea = :tada:

File: riff-raff/app/views/deploy/viewDeploy.scala.html [hunk 1]
```diff
@@ -30,9 +30,28 @@
...
+        </label>
+    </div>
+}
+@helper.form(action=routes.DeployController.processForm) {
+    <input type="hidden" name="project" value="@record.buildName"/>
+    <input type="hidden" name="build" value="@record.buildId"/>
+    <input type="hidden" name="stage" value="@record.stage.name"/>
+    <input type="hidden" name="recipe" value="@record.recipe.name"/>
+
+    @record.stacks.zipWithIndex.map { stack =>
```
Comment (adamnfish) [inline riff-raff/app/views/deploy/viewDeploy.scala.html]: would this be better written as something like:

```scala
  @record.stacks.zipWithIndex.map { case (index, stack) =>
    <input type="hidden" name="stacks[@index]" value="@{stack.name}"/>
```

**Comment by sihil** [comment]
I would recommend having a confirmation page. Conveniently this exists (possibly needs TLC) at https://github.com/guardian/riff-raff/blob/master/riff-raff/app/controllers/DeployController.scala#L194-L202

**Comment by jennysivapalan** [comment]
@sihil ok I can work on that.

**Comment by philwills** [comment]
There's various cruft we should remove in the post GC2 world. Specific hosts probably fit into that category.

**Comment by sihil** [comment]
I looked at doing that and then realised that riff-raff's own deploy still uses the specific host stuff.

**Comment by philwills** [comment]
🙀

**Comment by jennysivapalan** [comment]
Ok the Deploy Again button now goes to a confirmation page:

screen shot 2016-06-09 at 16 38 38

**Comment by sihil** [comment]
Nice!

On Thu, 9 Jun 2016 16:40 Jenny Sivapalan, <notifications@github.com> wrote:

> Ok the Deploy Again button now goes to a confirmation page:
>
> [image: screen shot 2016-06-09 at 16 38 38]
> <https://cloud.githubusercontent.com/assets/944375/15935869/cfa82154-2e60-11e6-9e5c-19ec87bd126d.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/guardian/riff-raff/pull/299#issuecomment-224935624>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABLd8p4VKMlwoh1GMb6OxZZWVzsyHBtKks5qKDP0gaJpZM4IombP>
> .
>

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

File: riff-raff/app/views/deploy/viewDeploy.scala.html [hunk 2]
```diff
@@ -32,9 +32,28 @@
...
-    <input id="verbose-checkbox" @if(verbose){checked="checked"} type="checkbox"/> Show verbose log lines
-</label>}
+@if(!record.isSummarised){
+    <div class="clearfix">
+        <label class="checkbox pull-right checkbox-fix-top">
+            <input id="verbose-checkbox" @if(verbose){checked="checked"} type="checkbox"/> Show verbose log lines
+        </label>
+    </div>
+}
+@helper.form(action=routes.DeployController.deployAgain) {
```
Comment (adamnfish) [inline riff-raff/app/views/deploy/viewDeploy.scala.html]: This is missing its CSRF field I think?

Comment (adamnfish) [inline riff-raff/app/views/deploy/viewDeploy.scala.html]: Ah don't worry. It's a GET with no side effects.

All good.

**Pull request merged**